Objects/obmalloc.c

```diff
@@ -2739,4 +2802,195 @@ _PyObject_DebugMallocStats(FILE *out)
+arena_map_get(block *p, int create)
+{
+    int i1 = MAP1_INDEX(p);
+    if (arena_map_root.ptrs[i1] == NULL) {
+        if (!create) {
+            return NULL;
+        }
+        arena_map2_t *n = PyMem_RawCalloc(1, sizeof(arena_map2_t));
+        if (n == NULL) {
+            return NULL;
+        }
+        arena_map_root.ptrs[i1] = n;
+        arena_map1_count++;
+    }
+    int i2 = MAP2_INDEX(p);
+    if (arena_map_root.ptrs[i1]->ptrs[i2] == NULL) {
+        if (!create) {
+            return NULL;
+        }
+        arena_map3_t *n = PyMem_RawCalloc(1, sizeof(arena_map3_t));
+        if (n == NULL) {
+            return NULL;
+        }
+        arena_map_root.ptrs[i1]->ptrs[i2] = n;
+        arena_map2_count++;
+    }
+    return arena_map_root.ptrs[i1]->ptrs[i2];
+}
+
+/* The radix tree only tracks arenas.  So, for 16 MiB arenas, we throw
+ * away 24 bits of the address.  That reduces the space requirement of
+ * the tree compared to similar radix tree page-map schemes.  In
+ * exchange for slashing the space requirement, it needs more
+ * computation to check an address.
+ *
+ * Tracking coverage is done by "ideal" arena address.  It is easier to
+ * explain in decimal so let's say that the arena size is 100 bytes.
+ * Then, ideal addresses are 100, 200, 300, etc.  For checking if a
+ * pointer address is inside an actual arena, we have to check two ideal
+ * arena addresses.  E.g. if pointer is 357, we need to check 200 and
+ * 300.  In the rare case that an arena is aligned in the ideal way
+ * (e.g. base address of arena is 200) then we only have to check one
+ * ideal address.
+ *
+ * The tree nodes for 200 and 300 both store the address of arena.
+ * There are two cases: the arena starts at a lower ideal arena and
+ * extends to this one, or the arena starts in this arena and extends to
+ * the next ideal arena.  The tail_lo and tail_hi members correspond to
+ * these two cases.
+ */
+
+
+/* mark or unmark addresses covered by arena */
+static int
+arena_map_mark_used(uintptr_t arena_base, int is_used)
+{
+    arena_map3_t *n_hi = arena_map_get((block *)arena_base, is_used);
+    if (n_hi == NULL) {
+        assert(is_used); /* otherwise node should already exist */
+        return 0; /* failed to allocate space for node */
+    }
+    int i3 = MAP3_INDEX((block *)arena_base);
+    int32_t tail = (int32_t)(arena_base & ARENA_MASK);
+    if (tail == 0) {
+        /* is ideal arena address */
+        n_hi->arenas[i3].tail_hi = is_used ? -1 : 0;
+    }
+    else {
+        /* arena_base address is not ideal (aligned to arena size) and
+         * so it potentially covers two MAP3 nodes.  Get the MAP3 node
+         * for the next arena.  Note that it might be in different MAP1
+         * and MAP2 nodes as well so we need to call arena_map_get()
+         * again (do the full tree traversal).
+         */
+        n_hi->arenas[i3].tail_hi = is_used ? tail : 0;
+        uintptr_t arena_base_next = arena_base + ARENA_SIZE;
```

Comment:
Alas, overflow is theoretically possible here.  Although it may happen only once in the next 50 years ☹️  If it does, there's no possibility for this arena to "spill over" into the next ideal arena address, so guarding the rest of this block with, e.g.,

`if (arena_base_next > arena_base) {  // no overflow`

should be fine.  Since this code is executed relatively rarely, I don't care about speed here.

Reply:
On second thought, if `arena_base` is a legit arena base address, then so is `arena_base + ARENA_SIZE - 1`, the last address in the arena.  So if 1 more than that overflows, it must overlow to 0, in which case `arena_base` was arena-aligned to begin with, and so would have taken the earlier "is ideal arena address" branch instead.  So I'd be happy here just to assert:

    assert(arena_base < arena_base_next); // no overflow